### DISCUSSION TITLE
Is there a way to change a node type from `node` to `node,app`?

### DISCUSSION THREAD

**Discussion by tukusejssirs** [gravitational/teleport#22429] [Q&A]
I have added a node dynamically using the `bash -c "$(curl ...install-node.sh)"` script. While the script has CLI options to add app name and URI, I presume that most Teleport users at that time don’t have any app set up (only a bare node), therefore they are useless. Therefore, the script sets the type of the node to `node` only (and not to `node,app`), which is saved on in the cluster. Therefore any change of this needs to be reflected in the cluster, therefore we need to follow [these docs](https://goteleport.com/docs/application-access/getting-started/#step-23-install-and-configure-teleport), however, it is cumbersome, as `tctl tokens add` can be run on `auth` type server (i.e. cluster) or we need to generate an indentity file using `tctl auth sign -o some_file`, which again can be done only on auth server.

Alternative to this is to not use the dynamic [addition of a node using CA pin](https://goteleport.com/docs/management/admin/adding-nodes/#step-23-join-your-node-to-the-cluster), but CA pins are invalidated when CA rotation is performed using `tctl auth rotate`.

All in all, neither of the solutions are sufficiently user-friendly, unfortunately.

---

Or is there a (simple) way to change the node type (from `node` to `node,app`)? Ideally, from the node itself?

I tried to define apps in `/etc/teleport.yaml`, but they were simply ignored (I presume because of the node type is `node`, not `node,app`). And when I run `teleport start --type node,app` (or even `--type app`), it failed with `ERROR: application name (--app-name) and URI (--app-uri) flags are both required to join application proxy to the cluster`, although they were defined in the config.

And when the app config is added to `/etc/teleport.yaml` and `teleport start` is executed, the log is flooded with errors like this:

```bash
App failed to establish connection to cluster: "$nodname" can not join the cluster with role "App", token expired or not found. pid:3180822.1 service/connect.go:119
```

---

IMHO, ideal solution for this issue is to  update the `install-node.sh` script to ask which node types/roles should be allowed, and even add a CLI option to configure them. Although the problem might be on the cluster side that it generates tokens for `node` only and not for `app`.

**Comment by webvictim** [comment]
As you've noticed, you can do this as follows:

- run `tctl tokens add --type=node,app` on your Teleport auth server and copy the outputted token
- paste the new token into `/etc/teleport.yaml` on your server (replace the existing one)
- add an `app_service` block to `/etc/teleport.yaml` on your server
- restart Teleport on your server

I know this is cumbersome, but we generally expect most people deploying Teleport on many servers to be installing the binaries and templating the config file themselves, rather than using the add server script.

As an aside, I believe if you add an application directly from the web UI, it does automatically add the SSH service as well as the application. It certainly used to, anyway. Even if you add a fake application which doesn't work, it'll still do most of the heavy lifting for you and you can change the `app_service` configuration to serve real apps later.

**Reply by tukusejssirs** [reply]
> we generally expect most people deploying Teleport on many servers to be installing the binaries and templating the config file themselves

Do you mean [creating invites](https://goteleport.com/docs/management/admin/adding-nodes/#step-23-join-your-node-to-the-cluster) on the `auth` server or using a [join token](https://goteleport.com/docs/server-access/getting-started/#create-a-join-token)?

---

Anyway, one thing I like in the dynamic node addition is the definition of a list of allowed system users to login via SSH on a particular node. How can I do that using a command?

---

If the dynamic no addition is a second-class citizen, why is it even propagated by Teleport? :thinking:

Now, I think that adding another step/configuration in the node/server addition process in the web UI to define what types should a new node have, is fairly simple to accomplish. Just like the selection all allowed node system users, the users should be able to choose the node type too. Should I open an issue for that?

**Reply by webvictim** [reply]
Go for it.

**Reply by tukusejssirs** [reply]
>> Should I open an issue for that?
>
> Go for it.

I presume your reaction is to this.

How about my other questions?

> > we generally expect most people deploying Teleport on many servers to be installing the binaries and templating the config file themselves
> 
> **[Question 1]**
>
> Do you mean [creating invites](https://goteleport.com/docs/management/admin/adding-nodes/#step-23-join-your-node-to-the-cluster) on the `auth` server or using a [join token](https://goteleport.com/docs/server-access/getting-started/#create-a-join-token)?
> 
> **[Question 2]**
>
> Anyway, one thing I like in the dynamic node addition is the definition of a list of allowed system users to login via SSH on a particular node. How can I do that using a command?
> 
> **[Question 3]**
>
> If the dynamic no addition is a second-class citizen, why is it even propagated by Teleport? 🤔

**Reply by webvictim** [reply]
> I presume your reaction is to this.

Yes.

> Do you mean [creating invites](https://goteleport.com/docs/management/admin/adding-nodes/#step-23-join-your-node-to-the-cluster) on the auth server or using a [join token](https://goteleport.com/docs/server-access/getting-started/#create-a-join-token)?

Either/or. When people install Teleport on many different servers, we expect that they will be using config management (like Ansible, Puppet, Chef, Salt etc) to template the config file, along with some out-of-band process which creates join tokens (say a job running on a timer) that then installs Teleport and templates the config out to the servers so they can join.

> Anyway, one thing I like in the dynamic node addition is the definition of a list of allowed system users to login via SSH on a particular node. How can I do that using a command?

You can set this in a Teleport role, then grant the `allow-bob` role to all users you want to be able to use that login:

```yaml
kind: role
metadata:
  name: allow-bob
spec:
  allow:
    logins:
    - bob
   node_labels:
     '*': '*'
```

Alternatively, you can do it using the `logins:` trait on a given Teleport user:

`tctl users update --set-logins bob teleport-user`

This essentially grants the given Teleport user permission for the `bob` login, like this:

```yaml
kind: user
metadata:
  name: teleport-user
spec:
  roles:
  - access
  traits:
    logins:
    - bob
```

> If the dynamic no addition is a second-class citizen, why is it even propagated by Teleport? 🤔

I don't completely understand this question. You mean the discover flow in the web UI which gives you the `curl | bash` type script, right?

The original intention of this script was essentially to allow people trying out Teleport for the first time to easily be able to add a node to their cluster for testing. It wasn't ever really designed to be run across a whole fleet of servers to set up Teleport - although it will work for that purpose, assuming that you only want to configure SSH _or_ apps _or_ databases etc. When you get into the state where you're installing Teleport for production use and want multiple services running, plus Teleport to be updated to new versions, you're better off building a process to automate the installation and config in your own environment.
